Store datasets used for scheduling on the timetable, not the DAG

ashb opened the pull request:
And as part of this, "tidy up" the API of BaseSerialization by making (de)serialize a public method.


This was mostly driven by a dislike of the attribute name "dataset_triggers" on the DAG (specifically the "triggers" part, as it is nothing to do with the existing "Trigger" concept in Airflow)

ashb committed:
Store datasets used for scheduling on the timetable, not the DAG

And as part of this, "tidy up" the API of BaseSerialization by making
(de)serialize a public method.

ashb committed:
fixup! Store datasets used for scheduling on the timetable, not the DAG

dstandish commented on airflow/models/dag.py:
```diff
@@ -2680,7 +2680,9 @@ class InletRef(NamedTuple):
         for dag in dags:
-            for dataset in dag.dataset_triggers:
+            if not isinstance(dag.timetable, DatasetTriggeredTimetable):
+                continue
```
i think exiting here means you don't iterate through outlets to get outlet datasets

ashb replied:
Whoops yep.

uranusjr committed:
Merge branch 'main' into datasets-on-timetable-not-dag

uranusjr committed:
Raise 'datasets' to base timetable class

This avoids many polymorphism breakages and allows for potential future
features (combine dataset-triggering and schedule-triggering in one
timetable).
